aiosmtpd/smtp.py

```diff
@@ -723,18 +728,34 @@ async def smtp_AUTH(self, arg: str) -> None:
             if status is not None:  # pragma: no branch
                 await self.push(status)
 
-    async def _auth_interact(self, server_message) -> _TriStateType:
-        blob: bytes
-        await self.push(server_message)
+    async def challenge_auth(
+            self,
+            challenge: AnyStr,
+            encode_to_b64: bool = True,
+    ) -> Union[_Missing, bytes]:
+        """
+        Send challenge during authentication. "334 " will be prefixed, so do NOT
+        put "334 " at start of server_message.
+
+        :param challenge: Challenge to send to client. Must be ASCII-encodable
+        :param encode_to_b64: If true, then perform Base64 encoding on challenge
+        :return: Response from client, or MISSING
+        """
+        challenge = (
+            challenge.encode("ascii") if isinstance(challenge, str) else challenge
+        )
+        assert isinstance(challenge, bytes)
+        # Trailing space is MANDATORY even if server_message is empty.
+        # See https://tools.ietf.org/html/rfc4954#page-4 ¶ 5
+        await self.push(
+            b"334 " + (b64encode(challenge) if encode_to_b64 else challenge)
+        )
         line = await self._reader.readline()
-        blob = line.strip()
-        # '=' and '*' handling are in accordance with RFC4954
-        if blob == b"=":
-            log.debug("%r responded with '='", self.session.peer)
-            return None
```

Comment:
This is my great mistake; I misunderstood RFC4954.

A sole equal sign `=` is supposed to be Base64-decoded _first_, and from there we get the empty data `b""`